### PR TITLE
ENH: Optimize polyutils as_series

### PR DESCRIPTION
**Pull request by eendebakpt**
Optimize `numpy.polynomial.polyutils.as_series` for the common case (arrays which are not of object dtype).

Benchmark:
```
import timeit
import numpy as np
from numpy.polynomial.polyutils import as_series
x=np.array([1.])
y=np.array([1.], dtype=object)

number=1_000_000
for alist in [ [x,], [x, x], [x, y]]:
    dt = timeit.timeit('as_series(alist, trim=True)', globals=globals(), number=number)
    print(f'as_series({alist}): {1e6*dt/number:.3f} [us]')
```

On main:
```
as_series([array([1.])]): 2.394 [us]
as_series([array([1.]), array([1.])]): 3.820 [us]
as_series([array([1.]), array([1.0], dtype=object)]): 3.349 [us]
```

This PR:
``` 
as_series([array([1.])]): 1.537 [us]
as_series([array([1.]), array([1.])]): 2.736 [us]
as_series([array([1.]), array([1.0], dtype=object)]): 3.190 [us]
```

### CODE REVIEW

**Comment by mattip** [comment]
What happens for larger array?

**Comment by eendebakpt** [comment]
> What happens for larger array?

The main computation time is in overhead and checking the dtypes and shapes, the actual computation with the data is small. Here is a benchmark for larger arrays:
```
Main:
as_series(['array(float64, shape (1000,))']): 2.241 [us]
as_series(['array(float64, shape (1000,))', 'array(float64, shape (1000,))']): 3.460 [us]
as_series(['array(float64, shape (1000,))', 'array(object, shape (300,))']): 13.872 [us]

PR:
as_series(['array(float64, shape (1000,))']): 1.525 [us]
as_series(['array(float64, shape (1000,))', 'array(float64, shape (1000,))']): 2.598 [us]
as_series(['array(float64, shape (1000,))', 'array(object, shape (300,))']): 14.192 [us]
```
The `as_series` is used for coefficients of polynomials. I suspect the arrays will not be much larger than 1000.

File: numpy/polynomial/polyutils.py [hunk 1]
```diff
@@ -118,25 +118,28 @@ def as_series(alist, trim=True):
...
-    if any(a.ndim != 1 for a in arrays):
-        raise ValueError("Coefficient array is not 1-d")
+        if a.ndim != 1:
+            raise ValueError("Coefficient array is not 1-d")
     if trim:
         arrays = [trimseq(a) for a in arrays]
 
-    if any(a.dtype == np.dtype(object) for a in arrays):
+    try:
+        dtype = np.common_type(*arrays)
```
Comment (seberg) [inline numpy/polynomial/polyutils.py:127]: Why can't we just use `result_type` and be done with it!?  Does this actually need any of the `common_type` shenanigans that don't really make sense?

**Review by seberg — APPROVED** [review]
I'll just put this in, thanks.  I don't think it is worse than the old code, maybe slightly clearer, since the try/except was always there.

It looks like code that shouldn't use `common_type` of course (and that would be faster since you should need none of the shenanigans, except the object copy/cast maybe -- I am not even sure about that part.)

So, it's a bit of a half improvement TBH, but considering that the PR was just inactive for so long, let me just put it in, thanks.

File: numpy/polynomial/polyutils.py (hunk 1 above)
Comment (eendebakpt) [reply numpy/polynomial/polyutils.py:127]: Maybe we can. There was another PR where I tried to replace `common_type` with `result_type`. It looked promising, but in the end it didn't work out. Maybe here it does work, I'll check at some point

**Pull request merged**